### PR TITLE
sql,backupccl: create IMPORT ROLLBACK job in online restore

### PR DESCRIPTION
**Pull request by stevendanna**
Tables that are offline and have a non-zero ImportEpoch are now restored to their offline state and an IMPORT
ROLLBACK job is synthesized.

The IMPORT ROLLBACK job uses an epoch-based predicate delete to
rollback the table and bring it back online.

Epic: none
Release note: None

### CODE REVIEW

File: pkg/ccl/backupccl/restore_job.go [hunk 1]
```diff
@@ -812,18 +812,31 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
...
-		return false, nil
+		return false
 	}
 
-	if table.GetInProgressImportStartTime() == 0 {
-		return false, nil
+	return table.GetInProgressImportStartTime() > 0
+}
+
+func backedUpDescriptorWithImportEpoch(desc catalog.Descriptor) bool {
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_job.go:824]: nit: could add a comment here? if a desc an importEpoch, it just means that an import was run at some point on table.

File: pkg/ccl/backupccl/restore_job.go [hunk 2]
```diff
@@ -886,22 +899,32 @@ func createImportingDescriptors(
...
 				offlineSchemas[schema.GetID()] = struct{}{}
 			}
 
-			if hasInProgressImportInto, err := backedUpDescriptorWithInProgressImportInto(ctx, p, desc); err != nil {
-				return nil, nil, nil, err
-			} else if hasInProgressImportInto && details.ExperimentalOnline {
-				return nil, nil, nil, errors.Newf("table %s (id %d) in restoring backup has an in-progress import, but online restore cannot be run on a table with an in progress import", desc.GetName(), desc.GetID())
-			} else if !hasInProgressImportInto {
+			if backedUpDescriptorWithInProgressImportInto(desc) {
+				if details.ExperimentalOnline && !backedUpDescriptorWithImportEpoch(desc) {
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_job.go:924]: nit: could we scrap `backedUpDescriptorWithImportEpoch` and just use `!epochBasedinProgressImport()`?

File: pkg/sql/importer/rollback_job.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		return txn.KV().Run(ctx, b)
+	})
+}
+
+func (*importRollbackResumer) CollectProfile(context.Context, interface{}) error {
+	return nil
+}
+
+func (r *importRollbackResumer) OnFailOrCancel(context.Context, interface{}, error) error {
+	return errors.AssertionFailedf("OnFailOrCancel called on non-cancellable job %d", r.job.ID())
```
Comment (msbutler) [inline pkg/sql/importer/rollback_job.go:125]: i may have missed this somewhere: do you have to flag in job creation or in the job constructor that this job is non-cancellable?

**Review by msbutler — COMMENTED** [review]
This is great! I gave a pass at the non-test code. It looks like your test failed in CI, so I'll wait to look at it once it's green.

File: pkg/sql/importer/rollback_job.go (hunk 3 above)
Comment (stevendanna) [reply pkg/sql/importer/rollback_job.go:125]: Good catch, added the bool to the job record.

**Comment by stevendanna** [comment]
> It looks like your test failed in CI, so I'll wait to look at it once it's green.

Yeah, I semantically merged-conflicted with myself.

File: pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports [hunk 4]
```diff
@@ -1,35 +1,66 @@
...
 job paused at pausepoint
 
-
 exec-sql
 BACKUP INTO 'nodelocal://1/cluster/';
 ----
 
-
-new-cluster name=s2 share-io-dir=s1 allow-implicit-access disable-tenant
+new-cluster name=s2 allow-implicit-access disable-tenant
```
Comment (msbutler) [inline pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports:47]: nit: no need to use a new cluster

File: pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports [hunk 5]
```diff
@@ -1,35 +1,66 @@
...
 exec-sql
-SET CLUSTER SETTING jobs.debug.pausepoints = 'import.after_ingest';
+EXPORT INTO CSV 'nodelocal://1/export1/' FROM SELECT * FROM baz;
+----
+NOTICE: EXPORT is not the recommended way to move data out of CockroachDB and may be deprecated in the future. Please consider exporting data with changefeeds instead: https://www.cockroachlabs.com/docs/stable/export-data-with-changefeeds
+
+
+
+exec-sql
+SET CLUSTER SETTING bulkio.backup.elide_common_prefix.enabled = false;
```
Comment (msbutler) [inline pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports:27]: why did you set this?

**Review by msbutler — APPROVED** [review]
Looks good! just had one question about the unit test

File: pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports (hunk 5 above)
Comment (stevendanna) [reply pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports]: This test was originally written at a point where getting a database require working with this option in place simply didn't work. I'd have to put together exactly when on master when that was to remember. But we can remove it now.

Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/online-restore-in-progress-imports]: Sounds good, thanks!

**Comment by stevendanna** [comment]
bors r=msbutler

**Comment by stevendanna** [comment]
bors r=msbutler

**Pull request merged**